msgpack/decode.ts

```diff
@@ -0,0 +1,190 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+export function decode(uint8: Uint8Array) {
+  const { value, consumed } = decodeSlice(uint8);
+
+  if (consumed < uint8.length) {
+    throw new EvalError("Messagepack decode did not consume whole array");
+  }
+
+  return value;
+}
+
+function decodeString(uint8: Uint8Array, size: number) {
+  return decoder.decode(uint8.subarray(0, size));
+}
+
+function decodeArray(uint8: Uint8Array, size: number) {
+  let total = 0;
+  const arr: unknown[] = [];
+
+  for (let i = 0; i < size; i++) {
+    const { value, consumed } = decodeSlice(uint8);
+    arr.push(value);
+    uint8 = uint8.subarray(consumed);
+    total += consumed;
+  }
+
+  return { value: arr, consumed: total };
+}
+
+function decodeMap(uint8: Uint8Array, size: number) {
+  let total = 0;
+  const map: Record<number | string, unknown> = {};
+
+  for (let i = 0; i < size; i++) {
+    const { value: key, consumed: keyConsumed } = decodeSlice(uint8);
+    uint8 = uint8.subarray(keyConsumed);
+    total += keyConsumed;
+    const { value, consumed: valueConsumed } = decodeSlice(uint8);
+    uint8 = uint8.subarray(valueConsumed);
+    total += valueConsumed;
+
+    if (typeof key !== "number" && typeof key !== "string") {
+      throw new EvalError(
+        "Messagepack decode came across an invalid type for a key of a map",
+      );
+    }
+
+    map[key] = value;
+  }
+
+  return { value: map, consumed: total };
+}
+
+const decoder = new TextDecoder();
+
+/**
+ * Given a uint8array which contains a msgpack object,
+ * return the value of the object as well as how many bytes
+ * were consumed in obtaining this object
+ */
+function decodeSlice(uint8: Uint8Array): { value: unknown; consumed: number } {
+  const dataView = new DataView(uint8.buffer);
```

Comment:
**issue:** This `DataView` is not used by all the match arms so fairly often this allocation / creation is wasted. But more than that, creating a `DataView` from an `ArrayBuffer` points it to the beginning of the buffer. All of the `getX` calls then read a value from the second byte of the buffer onwards, not from the second byte of the uint8 slice onwards.

It might be a good idea to create the `DataView` once and pass it through the call stack as a parameter? Then for reading you'd use the `uint8` slice's byte offset + 1.

Reply:
Makes sense to me to just have one big `DataView`. Melds especially well with the other change you suggested so I'll probably put both of those together.